File: src/test/scala/gameover/fwk/ai/impl/AStarTest.scala
package gameover.fwk.ai.impl

import com.badlogic.gdx.math.{Vector2, GridPoint2}
import gameover.fwk.ai.AStar
import gameover.fwk.libgdx.collection.GdxArray
import org.scalatest._

/**
 * Test on A*.
 * @author olmartin@expedia.com
 * @since 2015-08-09
 */
class AStarTest extends FlatSpec with MapForTest with SpriteForTest with GivenWhenThen {

  println(collisionDetector)

  "Calling A* with standard pathfinding" must "return an empty list of points if the target is the same" in {
    Given("a collision detector on a map")
    val pathFinding: AStar = new AStar(new ManhattanHComputeStrategy)

    When("looking for a path from a point to the same point")
    val x = 0.5f
    val y = 0.5f
    val path: GdxArray[GridPoint2] = pathFinding.findPath(x, y, x, y, findClosestPoint = false, collisionDetector)

    Then("the path return should be empty")
    assert(path.isEmpty)
  }

  it must "return null if path can't be found" in {
    Given("a collision detector on a map")
    val pathFinding: AStar = new AStar(new ManhattanHComputeStrategy)

    When("looking for a path from a point to a point not reachable")
    val path: GdxArray[GridPoint2] = pathFinding.findPath(0.5f, 0.5f, 0.5f, 4.5f, findClosestPoint = false, collisionDetector)

    Then("the path return should be null")
    assert(path === null)
  }

  it must "return a path to closest point if path can't be found but closest point flag was set" in {
    Given("a collision detector on a map")
    val pathFinding: AStar = new AStar(new ManhattanHComputeStrategy)

    When("looking for a path from a point to a point not reachable or the closest point")
    val path: GdxArray[GridPoint2] = pathFinding.findPath(0.5f, 0.5f, 0.5f, 4.5f, findClosestPoint = true, collisionDetector)

    Then("a path is returned to the closest point")
    assert(path !== null)
    assert(path nonEmpty)
    assert(path.size === 1)
    val nearestPoint: GridPoint2 = path.last
    assert(nearestPoint === new GridPoint2(0, 3))
  }

  it must "return a path to the point if point is reachable" in {
    Given("a collision detector on a map")
    val pathFinding: AStar = new AStar(new ManhattanHComputeStrategy)

    When("looking for a path from a point to a point not reachable or the closest point")
    val path: GdxArray[GridPoint2] = pathFinding.findPath(2.5f, 4.5f, 5.5f, 4.5f, findClosestPoint = false, collisionDetector)

    Then("a path is returned to the closest point")
    assert(path !== null)
    assert(path nonEmpty)
    assert(path.size > 1)
    val lastPoint: GridPoint2 = path.last
    assert(lastPoint === new GridPoint2(5, 4))
  }

  "Calling A* with smooth pathfinding" must "return a smooth path to the point if point is reachable" in {
    Given("a collision detector on a map")
    val pathFinding: AStar = new AStar(new ManhattanHComputeStrategy)
    And("a sprite on the map")
    sprite.setPosition(2.5f, 4.5f)

    When("looking for a path from a point to a point not reachable or the closest point")
    val path: GdxArray[Vector2] = pathFinding.findSmoothPath(sprite, 5.5f, 4.5f, findClosestPoint = false, collisionDetector)

    Then("a path is returned to the closest point")
    assert(path !== null)
    assert(path nonEmpty)
    assert(path.size > 1)
    val lastPoint: Vector2 = path.last
    assert(lastPoint === new Vector2(5.5f, 4.5f))
  }
}

File: src/test/kotlin/gameover/fwk/ai/impl/AStarTest.kt
package gameover.fwk.ai.impl

import com.badlogic.gdx.math.GridPoint2
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import gameover.fwk.ai.AStar
import gameover.fwk.ai.CollisionState
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class AStarTest {

    private val collisionDetector = MapForTest.collisionDetector
    private val pathFinding = AStar(ManhattanHComputeStrategy())


    init {
        println(collisionDetector)
    }

    @Test
    fun `Calling A* with standard path finding must return an empty list of points if the target is the same`(){
        val x = 0.5f
        val y = 0.5f
        val path = pathFinding.findPath(x, y, x, y, false, collisionDetector)

        assertThat(path).isEmpty()
    }

    @Test
    fun `it must return null if path can't be found`(){
        val path = pathFinding.findPath(0.5f, 0.5f, 0.5f, 4.5f, false, collisionDetector)

        assertThat(path).isNull()
    }

    @Test
    fun `it must return a path to closest point if path can't be found but closest point flag was set`(){
        val path = pathFinding.findPath(0.5f, 0.5f, 0.5f, 4.5f, true, collisionDetector)

        assertThat(path).isNotNull()
        assertThat(path).hasSize(1)
        val nearestPoint = path?.lastOrNull()!!
        assertThat(nearestPoint).isEqualTo(GridPoint2(0, 3))
    }

    @Test
    fun `it must return a path to the point if point is reachable - 1st use case`(){
        val origin = Vector2(2.5f, 4.5f)
        val target = Vector2(5.5f, 4.5f)
        val path = pathFinding.findPath(origin.x, origin.y, target.x, target.y, false, collisionDetector)

        assertThat(path).isNotNull()
        assertThat(path).hasSizeGreaterThan(1)
        val lastPoint = path?.lastOrNull()!!
        assertThat(lastPoint).isEqualTo(GridPoint2(5, 4))
        for (p in path.iterator()) {
            assertThat(collisionDetector.checkPosition(p.x.toFloat(), p.y.toFloat()))
                    .describedAs("Check if (${p.x}, ${p.y}) is valid")
                    .isEqualTo(CollisionState.Empty)
        }
        print("path from $origin to $target: $path")
    }

    @Test
    fun `it must return a path to the point if point is reachable - 2nd use case`(){
        val origin = Vector2(5.5f, 4.5f)
        val target = Vector2(2.5f, 4.5f)
        val path = pathFinding.findPath(origin.x, origin.y, target.x, target.y, false, collisionDetector)

        assertThat(path).isNotNull()
        assertThat(path).hasSizeGreaterThan(1)
        val lastPoint = path?.lastOrNull()!!
        assertThat(lastPoint).isEqualTo(GridPoint2(2, 4))
        for (p in path.iterator()) {
            assertThat(collisionDetector.checkPosition(p.x.toFloat(), p.y.toFloat()))
                    .describedAs("Check if (${p.x}, ${p.y}) is valid")
                    .isEqualTo(CollisionState.Empty)
        }
        print("path from $origin to $target: $path")
    }

    @Test
    fun `it must return a path to the point if point is reachable - 3rd use case`(){
        val origin = Vector2(4.5f, 3.5f)
        val target = Vector2(9.5f, 3.5f)
        val path = pathFinding.findPath(origin.x, origin.y, target.x, target.y, false, collisionDetector)

        assertThat(path).isNotNull()
        assertThat(path).hasSizeGreaterThan(1)
        val lastPoint = path?.lastOrNull()!!
        assertThat(lastPoint).isEqualTo(GridPoint2(9, 3))
        for (p in path.iterator()) {
            assertThat(collisionDetector.checkPosition(p.x.toFloat(), p.y.toFloat()))
                    .describedAs("Check if (${p.x}, ${p.y}) is valid")
                    .isEqualTo(CollisionState.Empty)
        }
        print("path from $origin to $target: $path")
    }

    @Test
    fun `it must return a path to the point if point is reachable - 4th use case`(){
        val origin = Vector2(4.5f, 3.5f)
        val target = Vector2(9.5f, 1.5f)
        val path = pathFinding.findPath(origin.x, origin.y, target.x, target.y, false, collisionDetector)

        assertThat(path).isNotNull()
        assertThat(path).hasSizeGreaterThan(1)
        val lastPoint = path?.lastOrNull()!!
        assertThat(lastPoint).isEqualTo(GridPoint2(9, 1))
        for (p in path.iterator()) {
            assertThat(collisionDetector.checkPosition(p.x.toFloat(), p.y.toFloat()))
                    .describedAs("Check if (${p.x}, ${p.y}) is valid")
                    .isEqualTo(CollisionState.Empty)
        }
        print("path from $origin to $target: $path")
    }

    @Test
    fun `Calling A* with smooth path finding must return a smooth path to the point if point is reachable`(){
        val sprite = Rectangle(2.5f, 4.5f, 0.2f, 0.2f)

        val path = pathFinding.findSmoothPath(sprite, 5.5f, 4.5f, false, collisionDetector)

        assertThat(path).isNotNull()
        assertThat(path).hasSizeGreaterThan(1)
        val lastPoint = path?.lastOrNull()!!
        assertThat(lastPoint).isEqualTo(Vector2(5.5f, 4.5f))
        for (p in path.iterator()) {
            assertThat(collisionDetector.checkPosition(p.x, p.y))
                    .describedAs("Check if (${p.x}, ${p.y}) is valid")
                    .isEqualTo(CollisionState.Empty)
        }
    }
}